flutter-idea/src/io/flutter/samples/FlutterSampleNotificationProvider.java

```diff
@@ -6,87 +6,98 @@
 
   public FlutterSampleNotificationProvider(@NotNull Project project) {
     this.project = project;
   }
 
-  @NotNull
-  @Override
-  public Key<EditorNotificationPanel> getKey() {
-    return KEY;
-  }
-
   @Nullable
   @Override
-  public EditorNotificationPanel createNotificationPanel(@NotNull VirtualFile file,
-                                                         @NotNull FileEditor fileEditor,
-                                                         @NotNull Project project) {
-    if (!(fileEditor instanceof TextEditor textEditor)) {
-      return null;
-    }
-
+  public Function<? super @NotNull FileEditor, ? extends @Nullable JComponent> collectNotificationData(@NotNull Project project,
+                                                                                                       @NotNull VirtualFile file) {
     final FlutterSdk sdk = FlutterSdk.getFlutterSdk(project);
     if (sdk == null) {
       return null;
     }
 
     final String flutterPackagePath = sdk.getHomePath() + "/packages/flutter/lib/src/";
-    final String filePath = file.getPath();
 
     // Only show for files in the flutter sdk.
+    final String filePath = file.getPath();
     if (!filePath.startsWith(flutterPackagePath)) {
       return null;
     }
 
+    return fileEditor -> createPanelForSamples(fileEditor, project, file, filePath, sdk, flutterPackagePath);
+  }
+
+  @Nullable
+  private EditorNotificationPanel createPanelForSamples(@NotNull FileEditor fileEditor,
+                                                        @NotNull Project project,
+                                                        @NotNull VirtualFile file,
+                                                        @NotNull String filePath,
+                                                        @NotNull FlutterSdk sdk,
+                                                        @NotNull String flutterPackagePath) {
+    if (!(fileEditor instanceof TextEditor textEditor)) {
+      return null;
+    }
+
     final Editor editor = textEditor.getEditor();
     final Document document = editor.getDocument();
+    final PsiDocumentManager psiDocumentManager = PsiDocumentManager.getInstance(project);
+    if (psiDocumentManager == null) {
+      return null;
+    }
 
-    final PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(document);
+    final PsiFile psiFile = psiDocumentManager.getPsiFile(document);
     if (psiFile == null || !psiFile.isValid()) {
       return null;
     }
 
     // Run the code to query the document in a read action.
     final List<FlutterSample> samples = ApplicationManager.getApplication().
       runReadAction((Computable<List<FlutterSample>>)() -> {
-        //noinspection CodeBlock2Expr
         return getSamplesFromDoc(flutterPackagePath, document, filePath);
       });
 
-    return samples.isEmpty() ? null : new FlutterSampleActionsPanel(samples);
+    if (samples != null && !samples.isEmpty()) {
```

Comment:
Can `samples` be null here or can you skip this check? (It looks like `getSamplesFromDoc` at worst returns an empty list?)

Reply:
Ah! So maybe this is new Java. I would have expected `ApplicationManager.getApplication(). runReadAction` to throw an NPE if `getApplication` returns null, but cool!